### ISSUE TITLE
Pass result of run() to afterRun() in yii\base\Action?

### ISSUE THREAD

**Issue by egml** [yiisoft/yii2#13386]
It seems reasonable when you inherit some base action class and want to further process the result or change behavior of `afterRun()` depending on the result. But are there any caveats of passing whole result every action run? Or any other methods? Right now inheriting `yii\base\Action` and duplicating `runWithParams()` doesn't seem ok.
Some distantly related issue: https://github.com/yiisoft/yii2/issues/2549.

**Comment by cebe** [comment]
Sorry I read your message three times but I do not get what you are asking :) Could you rephrase or explain in more detail please?

**Comment by egml** [comment]
Sorry) I'm talking about this piece of code here https://github.com/yiisoft/yii2/blob/master/framework/base/Action.php#L94-L95
```php
public function runWithParams($params)
{
	...
	if ($this->beforeRun()) {
		$result = call_user_func_array([$this, 'run'], $args);
		$this->afterRun();
		return $result;
	}
	...
}
```
Wouldn't it be right to pass `$result` to `afterRun()` before returning it? Like this:
```php
	...
	if ($this->beforeRun()) {
		$result = call_user_func_array([$this, 'run'], $args);
		$this->afterRun($result);
		return $result;
	}
	...
```
To be able to analyze the result of the action in `afterRun()` and change `afterRun()`'s execution accordingly. Right now I came up with a solution where I pass stuff through action class parameters. But it just seems not right. Or maybe I'm wrong, that was the base question) Is it worth making a PR or even talking about?

An example: action class which sends some specific [EventSource](https://developer.mozilla.org/en-US/docs/Web/API/EventSource) stuff to the browser before and after each action.
```php
class Operation extends \yii\base\Action
{
	...
	protected function beforeRun()
	{
		\Yii::$app->getLog()->targets['operation']->rotateFiles();
		$this->log('Operation begin [' . date('Y-m-d H:i:s') . ']');
		EventSource::begin();
		return true;
	}

	protected function afterRun()
	{
		... this is where it would be nice to have action run result to change $data ...
		EventSource::end($data);
		$this->log('Operation end [' . date('Y-m-d H:i:s') . ']');
	}
	...
}
```

**Comment by samdark** [comment]
Sounds reasonable for 2.1 but I'd go a bit further by passing result through allowing to post-process it:

```php
return $this->afterRun($result);
```

Can't be done in 2.0 because of method signature change.

**Comment by lubobill1990** [comment]
+1
I think it's reasonable to pass the result to afterRun

With out the result of run() to be passed to afterRun, afterRun can do very few things.
